client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -174,22 +187,46 @@ public AsyncIterator(Iterator<T> dataIterator, StatementClient client)
                 }
                 finally {
                     semaphore.release();
+                    finished = true;
                 }
-            }, executorService);
+            });
         }
 
         public void cancel()
         {
             future.cancel(true);
+            cancelled = true;
```

Comment:
`cancelled = true;` should be done before sending interrupt (`future.cancel(true)`), so that we have the guarantee that the woken up thread sees the right value of the cancellation flag

Reply:
It doesn't matter. Interruptions are more effiencient than loop check if interruptions are well handled. Because iteration may waiting for the result data to be produced.

Reply:
> . Interruptions are more effiencient than loop check if interruptions are well handled.

Yes

We want to have two mechanisms for the case where interruptions are not well handled.
In such cases, setting a flag first is important.

Reply:
Applied myself